File: src/main/scala-2.11/com/lei/leetcode/P46/Solution.java
package com.lei.leetcode.P46;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    private List<List<Integer>> mix(List<List<Integer>> data, int val) {
        List<List<Integer>> r = new ArrayList<>();
        for (int i = 0; i < data.size(); i++) {
            for (int j = 0; j <= data.get(i).size(); j++) {
                List<Integer> round = new ArrayList<>(data.get(i));
                round.add(j, val);
                r.add(round);
            }
        }
        return r;
    }

    public List<List<Integer>> permute(int[] nums) {
        if (nums.length == 1) {
            ArrayList<List<Integer>> r = new ArrayList<>();
            r.add(Arrays.asList(nums[0]));
            return r;
        } else if (nums.length == 0) {
            return new ArrayList<>();
        } else {
            ArrayList<List<Integer>> result = new ArrayList<>();
            int[] d = Arrays.copyOfRange(nums, 1, nums.length);
            int elem = nums[0];
            List<List<Integer>> restPermute = permute(d);
            result.addAll(mix(restPermute, elem));
            return result;
        }
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        List<List<Integer>> l = s.permute(new int[]{1, 2, 3});
        for (List<Integer> ls : l) {
            for (int i : ls) {
                System.out.print(i + " ");
            }
            System.out.println();
        }
    }

}

File: src/main/scala-2.11/com/lei/leetcode/P46/Solution.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.permutation(res, [], nums)
        return res

    def permutation(self, res, prefix, nums):
        if len(prefix) == len(nums):
            res.append(prefix[:])
        else:
            for i in range(0, len(nums)):
                if nums[i] in prefix:
                    continue
                prefix.append(nums[i])
                self.permutation(res, prefix, nums)
                prefix.pop(len(prefix) - 1)


if __name__ == '__main__':
    s = Solution()
    print s.permute([1, 2, 3])
